numpy/core/tests/test_multiarray.py

```diff
@@ -5898,6 +5969,15 @@ def test_matmul_inplace():
         exec_ = getattr(builtins, "exec")
         assert_raises(TypeError, exec_, "a @= b", globals(), locals())
 
+    def test_matmul_axes():
+        a = np.arange(3*4*5).reshape(3, 4, 5)
+        c = np.matmul(a, a, axes=[(-2, -1), (-1, -2), (1, 2)])
+        assert c.shape == (3, 4, 4)
+        d = np.matmul(a, a, axes=[(-2, -1), (-1, -2), (0, 1)])
+        assert d.shape == (4, 4, 3)
+        e = np.swapaxes(d, 0, 2)
+        assert_array_equal(e, c)
+
```

Comment:
It is annoying, but we probably should test the  matrix times vector case as well (where of course only the axes of the matrices make any sense).

Reply:
added test